python/ray/remote_function.py

```diff
@@ -254,6 +265,17 @@ def _remote(self, args=None, kwargs=None, **task_options):
         worker = ray._private.worker.global_worker
         worker.check_connected()
 
+        # We cannot do this when the function is first defined, because we need
+        # ray.init() to have been called when this executes
+        if self._function_signature is None:
+            # Since variable assignment is atomic, we don't need to acquire the
+            # lock before checking _function_signature.
+            with self._inject_lock:
```

Comment:
actually can you do 


```suggestion
            with self._inject_lock:
                if self._function_signature is None:
                    ...
```

Otherwise, there's race condition between checking function signature and lock

Reply:
there shouldn't be a race because variable assignment is atomic, though?

Reply:
oops no data race but still should have checked again inside the with